core/embed/bootloader/main.c

```diff
@@ -388,47 +392,55 @@ int bootloader_main(void) {
   }
 #endif
 
-  // start the bootloader if no or broken firmware found ...
-  if (firmware_present != sectrue) {
+  // ... or if user touched the screen on start
+  // ... or we have stay_in_bootloader flag to force it
+  // ... or there is no valid firmware
+  if (touched || stay_in_bootloader == sectrue || firmware_present != sectrue) {
+    screen_t screen;
+    if (header_present == sectrue) {
+      ui_set_initial_setup(false);
+      screen = SCREEN_INTRO;
+    } else {
+      screen = SCREEN_WELCOME;
 #ifdef TREZOR_EMULATOR
-    // wait a bit so that the empty lock icon is visible
-    // (on a real device, we are waiting for touch init which takes longer)
-    hal_delay(400);
+      // wait a bit so that the empty lock icon is visible
+      // (on a real device, we are waiting for touch init which takes longer)
+      hal_delay(400);
 #endif
-    // ignore stay in bootloader
-    stay_in_bootloader = secfalse;
-    touched = false;
-
-    // show intro animation
-
-    ui_set_initial_setup(true);
 
-    ui_screen_welcome_model();
-    hal_delay(1000);
-    ui_screen_welcome();
+      // erase storage
+      ensure(flash_erase_sectors(STORAGE_SECTORS, STORAGE_SECTORS_COUNT, NULL),
+             NULL);
 
-    // erase storage
-    ensure(flash_erase_sectors(STORAGE_SECTORS, STORAGE_SECTORS_COUNT, NULL),
-           NULL);
-
-    // and start the usb loop
-    if (bootloader_usb_loop(NULL, NULL) != CONTINUE) {
-      return 1;
+      ui_set_initial_setup(true);
```

Comment:
please move these lines above the emulator ifdef, so that it's easier to read all that happens in the else branch

Reply:
in fact, there is probably an overall better place for this ifdef block. something like at the else-branch of the "wait for touch" sequence? or maybe even just at the top level right after showing the boot screen the first time. emulator users can stand to wait 400ms in every case, and it will make it possible for people to trigger go-to-bootloader for TR which does not have a wait-loop.

(speaking of, could we again move `screen_boot_empty` at the very start, before reading the headers?)

Reply:
moved in https://github.com/trezor/trezor-firmware/commit/23c3b2b8752d3772d2421b3ffe3e83478f7b81a8